src/daft-scan/src/scan_task_iters.rs

```diff
@@ -0,0 +1,78 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+
+use crate::{ScanTask, ScanTaskRef};
+
+type BoxScanTaskIter = Box<dyn Iterator<Item = DaftResult<ScanTaskRef>>>;
+
+/// Coalesces ScanTasks by their filesizes
+pub fn merge_by_sizes(scan_tasks: BoxScanTaskIter, target_filesize: usize) -> BoxScanTaskIter {
+    Box::new(MergeByFileSize {
+        iter: scan_tasks,
+        target_filesize,
+        buffer: std::default::Default::default(),
+    })
+}
+
+struct MergeByFileSize {
+    iter: BoxScanTaskIter,
+    target_filesize: usize,
+
+    // HACK: we use a Vec instead of IndexMap because PartitionSpec is not Hashable
+    // This might potentially be very slow. We should explore another option here?
+    // Could be easier to make `PartitionSpec: Hash` if we don't use a `Table` under the hood.
+    buffer: Vec<ScanTaskRef>,
+}
+
+impl Iterator for MergeByFileSize {
+    type Item = DaftResult<ScanTaskRef>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.iter.next() {
+            Some(Ok(child_item)) => {
+                // Try to find matches in the buffer to perform merging operations
+                for (idx, buffered_item) in self.buffer.iter().enumerate() {
```

Comment:
I dont think we would want to go with an algorithm that uses the whole window of ScanTask since that will change the order of data that is requested. I was more thinking like a folding style reduce that only merges neighboring ScanTasks that will preserve ordering.
```
for task in tasks:
  if can_merge:
    curr = merge(curr, task)
  else:
    yield curr
    curr = task
```

Reply:
makes sense -- changed to use a single element iterator instead